File: ruby/saddle-points/saddle_points.rb
class Matrix
  attr_reader :rows, :columns
  def initialize(str)
    @rows = str.each_line.map{ |row| row.split.map(&:to_i) }
    @columns = @rows.transpose
  end

  def saddle_points
    saddle_points = []

    @rows.each_with_index do |row, yi|
      row.each_with_index do |n, xi|
        if biggest_of_row?(row, n) && smallest_of_column?(@columns[xi], n)
          saddle_points.push([yi, xi])
        end
      end
    end
    saddle_points
  end

  private

  def biggest_of_row?(row, n)
    n == row.max
  end

  def smallest_of_column?(col, n)
    n == col.min
  end
end

File: javascript/saddle-points/saddle-points.js
export class Matrix {
  constructor(strInput) {
    this.rows = strInput.split(/\n/).map(str => str.split(' ').map(char => parseInt(char, 10)));
    this.columns = this.rows[0].map((col, i) => this.rows.map(row => row[i]));
    this.saddlePoints = [];
    this.findSaddlePoints();
  }

  findSaddlePoints() {
    this.rows.forEach((row, rowIndex) => {
      row.forEach((element, columnIndex) => {
        const largestOfRow = Math.max(...row);
        const smallestOfColumn = Math.min(...this.columns[columnIndex]);
        if (element === largestOfRow && element === smallestOfColumn) {
          this.saddlePoints.push([rowIndex, columnIndex]);
        }
      });
    });
  }
}
